ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/pathbased/bonsai/worldview/BlockAccessListStateRootHashCalculator.java

```diff
@@ -0,0 +1,142 @@
+      final ProtocolContext protocolContext, final BlockHeader blockHeader) {
+    final BlockHeader chainHeadHeader = protocolContext.getBlockchain().getChainHeadHeader();
+    if (!chainHeadHeader.getHash().equals(blockHeader.getParentHash())) {
+      throw new IllegalStateException("Chain head is not the parent of the processed block");
+    }
+    final BonsaiWorldState ws =
+        (BonsaiWorldState)
+            protocolContext
+                .getWorldStateArchive()
+                .getWorldState(
+                    WorldStateQueryParams.withBlockHeaderAndNoUpdateNodeHead(chainHeadHeader))
+                .orElseThrow();
+    ws.disableCacheMerkleTrieLoader();
+    return ws;
+  }
+
+  private static Hash accumulateAccessListAndComputeRoot(
+      final BonsaiWorldState worldState, final BlockAccessList blockAccessList) {
+    final BonsaiWorldStateUpdateAccumulator accumulator =
+        (BonsaiWorldStateUpdateAccumulator) worldState.getAccumulator();
+
+    for (AccountChanges accountChanges : blockAccessList.accountChanges()) {
+      final Address address = accountChanges.address();
+      final MutableAccount account = accumulator.getOrCreate(address);
+
+      final List<BalanceChange> balanceChanges = accountChanges.balanceChanges();
+      if (!balanceChanges.isEmpty()) {
+        final BalanceChange change = balanceChanges.get(balanceChanges.size() - 1);
+        account.setBalance(Wei.wrap(change.postBalance()));
+      }
+
+      final List<NonceChange> nonceChanges = accountChanges.nonceChanges();
+      if (!nonceChanges.isEmpty()) {
+        final NonceChange change = nonceChanges.get(nonceChanges.size() - 1);
+        account.setNonce(change.newNonce());
+      }
+
+      final List<CodeChange> codeChanges = accountChanges.codeChanges();
+      if (!codeChanges.isEmpty()) {
+        final CodeChange change = codeChanges.get(codeChanges.size() - 1);
+        account.setCode(change.newCode());
+      }
+
+      for (SlotChanges slotChanges : accountChanges.storageChanges()) {
+        final List<StorageChange> changes = slotChanges.changes();
+        if (!changes.isEmpty()) {
+          final StorageChange change = changes.get(changes.size() - 1);
+          final Optional<UInt256> maybeKey = slotChanges.slot().getSlotKey();
+          if (maybeKey.isPresent()) {
+            final UInt256 key = maybeKey.get();
+            final UInt256 value = change.newValue();
+            account.setStorageValue(key, value == null ? UInt256.ZERO : value);
+          }
+        }
+      }
+    }
+
+    accumulator.commit();
+    return worldState.calculateRootHash(Optional.empty(), accumulator);
+  }
+
+  public static CompletableFuture<Hash> computeStateRootFromBlockAccessListAsync(
+      final ProtocolContext protocolContext,
+      final BlockHeader blockHeader,
+      final BlockAccessList bal) {
+
+    final BonsaiWorldState ws = prepareWorldState(protocolContext, blockHeader);
```

Comment:
why not doing prepareWorldState in background ? I think it can impact performance to do it in the main thread

Reply:
Debugging leftover, my bad :+1: